examples/polypharmacy_gripnet/model.py

```diff
@@ -1,86 +1,131 @@
+class MultiRelaInnerProductDecoder(torch.nn.Module):
     """
+	Build `DistMult
+	<https://arxiv.org/abs/1412.6575>`_ factorization as GripNet decoder in PoSE dataset.
+	Copy-paste with slight modifications from https://github.com/NYXFLOWER/GripNet
+	"""
 
-    def __init__(self, in_dim, num_et):
+    def __init__(self, in_channels: int, num_edge_type: int):
         super(MultiRelaInnerProductDecoder, self).__init__()
-        self.num_et = num_et
-        self.in_dim = in_dim
-        self.weight = nn.Parameter(torch.Tensor(num_et, in_dim))
+        self.num_edge_type = num_edge_type
+        self.in_channels = in_channels
+        self.weight = torch.nn.Parameter(torch.Tensor(num_edge_type, in_channels))
 
         self.reset_parameters()
 
-    def forward(self, z, edge_index, edge_type, sigmoid=True):
+    def forward(self, x, edge_index: torch.Tensor, edge_type: torch.Tensor, sigmoid: bool = True) -> torch.Tensor:
         """
-        Args:
-            z: input node feature embeddings.
-            edge_index: edge index in COO format with shape [2, num_edges].
-            edge_type: The one-dimensional relation type/index for each target edge in edge_index.
-            sigmoid: use sigmoid function or not.
-        """
-        value = (z[edge_index[0]] * z[edge_index[1]] * self.weight[edge_type]).sum(dim=1)
+		Args:
+			x: input node feature embeddings.
+			edge_index: edge index in COO format with shape [2, num_edges].
+			edge_type: The one-dimensional relation type/index for each target edge in edge_index.
+			sigmoid: use sigmoid function or not.
+		"""
+        value = (x[edge_index[0]] * x[edge_index[1]] * self.weight[edge_type]).sum(dim=1)
         return torch.sigmoid(value) if sigmoid else value
 
     def reset_parameters(self):
-        self.weight.data.normal_(std=1 / np.sqrt(self.in_dim))
+        self.weight.data.normal_(std=1 / np.sqrt(self.in_channels))
 
+    def __repr__(self) -> str:
+        return "{}: DistMultLayer(in_channels={}, num_relations={})".format(
+            self.__class__.__name__, self.in_channels, self.num_edge_type
+        )
 
-class GripNet(nn.Module):
-    """
-    Build GripNet-DistMult (Encoder-Decoder) model for PoSE link prediction.
-    """
 
-    def __init__(
-        self,
-        gene_channels_list,
-        gd_channels_list,
-        drug_channels_list,
-        num_drug_nodes,
-        num_gene_nodes,
-        num_drug_edge_relations,
-    ):
-        """
-        Parameter meanings explained in kale.embed.gripnet module.
-        """
-        super(GripNet, self).__init__()
-        self.num_drug_nodes = num_drug_nodes
-        self.num_gene_nodes = num_gene_nodes
-        self.gn = TypicalGripNetEncoder(
-            gene_channels_list,
-            gd_channels_list,
-            drug_channels_list,
-            num_drug_nodes,
-            num_gene_nodes,
-            num_drug_edge_relations,
+class GripNetLinkPrediction(pl.LightningModule):
```

Comment:
This is highly reusable and should go into the core API, likely `kale.predict`, also under `decode`?